src/base/bittorrent/torrentinfo.cpp

```diff
@@ -213,24 +220,81 @@ QByteArray TorrentInfo::metadata() const
 
 QStringList TorrentInfo::filesForPiece(int pieceIndex) const
 {
-    if (pieceIndex < 0)
-        return QStringList();
+    // no checks here because fileIndicesForPiece() will return an empty list
+    QVector<int> fileIndices = fileIndicesForPiece(pieceIndex);
+
+    QStringList res;
+    res.reserve(fileIndices.size());
+    std::transform(fileIndices.begin(), fileIndices.end(), std::back_inserter(res), [this](int i) {
+        return this->filePath(i);
+    });
+
+    return res;
+}
+
+QVector<int> BitTorrent::TorrentInfo::fileIndicesForPiece(int pieceIndex) const
+{
+    if (!isValid() || (pieceIndex < 0) || (pieceIndex >= piecesCount()))
+        return QVector<int>();
 
     std::vector<libtorrent::file_slice> files(
         nativeInfo()->map_block(pieceIndex, 0, nativeInfo()->piece_size(pieceIndex)));
-    QStringList res;
-    for (const libtorrent::file_slice& s: files) {
-        res.append(filePath(s.file_index));
-    }
+    QVector<int> res;
+    res.reserve(files.size());
+    std::transform(files.begin(), files.end(), std::back_inserter(res), [](const libtorrent::file_slice &s) {
+        return s.file_index;
+    });
+
     return res;
 }
 
+QPair<int, int> BitTorrent::TorrentInfo::piecesForFile(const QString& file) const
+{
+    if (!isValid()) // if we do not check here the debug message will be printed, which would be not correct
+        return qMakePair(-1, -1);
+
+    int index = fileIndex(file);
+    if (index == -1) {
+        qDebug() << "Filename '" << file << "' was not found in torrent '" << name() << '\'';
```

Comment:
Minor. Do you remember QDebug stream output adds spaces so you will get `Filename ' file ' was not found in torrent ' torrent name '`.